Jack-in als live share guest not working (polylith repo)

In one of the last versions up to v2.0.257 this stopped working. I don't know which one exactly.

We're using polylith.

When trying to connect to the remote REPL as a live share guest, I'm presented with a list of project roots. The paths are correct, but different from local. Locally, the paths are absolute paths from the root of my file system. Over live share the paths are absolute from the root of the workspace. For example, locally I might have `/Users/my/my-project` and when connecting over live share that would show up as `/`.

I choose ‘/’, then I'm asked for the type, I choose ‘generic’, and then nothing…

@PEZ Yes that `await vsls.getApi()` resolves for me.

I have run this scenario in the debugger, the live share guest that is. A couple of observations.

## clojure-lsp error

This one happens right on joining the Live Share session, before trying to connect the REPL or something.

```
Clojure-lsp zip file downloaded to [/Users/stefanmm/Documents/Projects/3dparty/calva/clojure-lsp-native-macos-amd64.zip]()
src/lsp/download.ts:61
Unzipping file
src/lsp/download.ts:86
[emma-mono[/RunTestListener]()] Jest process "watch-tests" ended unexpectedly
/Users/stefanmm/.vscode/extensions/orta.vscode-jest-4.2.1/out/extension.js:2
Writing version file
src/lsp/download.ts:76
Starting clojure-lsp at [/Users/stefanmm/Documents/Projects/3dparty/calva/clojure-lsp]()
src/lsp/main.ts:403
clojure[/serverInfo/raw]() failed: Error: Internal error.
    at handleResponse ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:477:48]())
    at processMessageQueue ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:292:17]())
    at Immediate.<anonymous> ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:276:13]())
    at processImmediate ([/Users/stefanmm/Documents/Projects/3dparty/calva/lib/internal/timers.js:464:21]())
    at process.callbackTrampoline (node:internal[/async_hooks:130:17]()) {code: -32603, data: 'java.lang.RuntimeException: java.lang.reflect.…r.serverInfoRaw(server.clj:135)
	... 15 more
', stack: 'Error: Internal error.
    at handleResponse …Trampoline (node:internal[/async_hooks:130:17]())', message: 'Internal error.'}

arg1:Error: Internal error.\n    at handleResponse ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:477:48]())\n    at processMessageQueue ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:292:17]())\n    at Immediate.<anonymous> ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:276:13]())\n    at processImmediate (node:internal[/timers:464:21]())\n    at process.callbackTrampoline (node:internal[/async_hooks:130:17]()) {code: -32603, data: 'java.lang.RuntimeException: java.lang.reflect.…r.serverInfoRaw(server.clj:135)\n\t... 15 more\n', stack: 'Error: Internal error.\n    at handleResponse …Trampoline (node:internal[/async_hooks:130:17]())', message: 'Internal error.'}
code:-32603
data:'java.lang.RuntimeException: java.lang.reflect.InvocationTargetException\n\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)\n\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessa…lj:20)\n\tat clojure_lsp.feature.format$resolve_user_cljfmt_config.invoke(format.clj:19)\n\tat clojure_lsp.handlers$server_info$fn__34629.invoke(handlers.clj:262)\n\tat clojure_lsp.handlers$server_info.invokeStatic(handlers.clj:261)\n\tat clojure_lsp.handlers$server_info.invoke(handlers.clj:253)\n\tat clojure.lang.Var.invoke(Var.java:380)\n\tat clojure_lsp.handlers.ClojureLSPFeatureHandler.server_info_raw(handlers.clj:532)\n\tat clojure_lsp.server.ClojureLspServer.serverInfoRaw(server.clj:135)\n\t... 15 more...
message:'Internal error.'
stack:'Error: Internal error.\n    at handleResponse ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:477:48]())\n    at processMessageQueue ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:292:17]())\n    at Immediate.<anonymous> ([/Users/stefanmm/Documents/Projects/3dparty/calva/node_modules/vscode-jsonrpc/lib/common/connection.js:276:13]())\n    at processImmediate (node:internal[/timers:464:21]())\n    at process.callbackTrampoline (node:internal[/async_hooks:130:17]())'
[[Prototype]]:Error
<anonymous> @ [/Applications/Visual]() Studio Code.app[/Contents/Resources/app/out/bootstrap-fork.js:5:6]()
```

Screenshot:

<img width="1454" alt="Screenshot 2022-04-02 at 10 50 11" src="https://user-images.githubusercontent.com/4028554/161375411-f4642d7f-fcb5-403a-9388-24a4c040c6e6.png">

The call stack of the internal responseMessage.error:

```
"java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
  at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
  at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
  at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
  at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
  at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
  at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
  at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
  at java.util.concurrent.FutureTask.run(FutureTask.java:264)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(Thread.java:829)
  at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:597)
  at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:194)
Caused by: java.lang.reflect.InvocationTargetException
  at java.lang.reflect.Method.invoke(Method.java:566)
  at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
  ... 13 more
Caused by: java.lang.NullPointerException
  at clojure.string$starts_with_QMARK_.invokeStatic(string.clj:365)
  at clojure.string$starts_with_QMARK_.invoke(string.clj:361)
  at clojure_lsp.shared$uri__GT_filename.invokeStatic(shared.clj:161)
  at clojure_lsp.shared$uri__GT_filename.invoke(shared.clj:156)
  at clojure_lsp.feature.format$resolve_user_cljfmt_config.invokeStatic(format.clj:20)
  at clojure_lsp.feature.format$resolve_user_cljfmt_config.invoke(format.clj:19)
  at clojure_lsp.handlers$server_info$fn__34629.invoke(handlers.clj:262)
  at clojure_lsp.handlers$server_info.invokeStatic(handlers.clj:261)
  at clojure_lsp.handlers$server_info.invoke(handlers.clj:253)
  at clojure.lang.Var.invoke(Var.java:380)
  at clojure_lsp.handlers.ClojureLSPFeatureHandler.server_info_raw(handlers.clj:532)
  at clojure_lsp.server.ClojureLspServer.serverInfoRaw(server.clj:135)
  ... 15 more
"
```

So I guess clojure-lsp doesn't like live share? Well I already knew that I guess, but maybe good to solve at some point.

## Error when connecting REPL

This one is probably why the REPL connection is not working:

```
rejected promise not handled within 1 second: Unknown (FileSystemError) (FileSystemError): Error: EROFS: read-only file system, mkdir '[/.calva]()'
/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:104
stack trace: Unknown (FileSystemError) (FileSystemError): Error: EROFS: read-only file system, mkdir '[/.calva]()'
    at Function._handleError ([/Applications/Visual]() Studio Code.app[/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:85:88571]())
    at Object.createDirectory ([/Applications/Visual]() Studio Code.app[/Contents/Resources/app/out/vs/workbench/api/node/extensionHostProcess.js:85:86821]())
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal[/process/task_queues:96:5]())
```

Screenshot:

<img width="1450" alt="Screenshot 2022-04-02 at 10 51 28" src="https://user-images.githubusercontent.com/4028554/161375507-b4bfa7c2-2605-411b-b28f-c8512ac11284.png">

I'm not sure how to proceed. I don't know why this `mkdir` is being called and what triggers it. It should not happen of course, because it should be the remote workspace's `.calva` folder that is being used (and already exists) it seems to me.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,8 @@ Changes to Calva.
 
 ## [Unreleased]
 
+- Fix: [Jack-in as live share guest not working](https://github.com/BetterThanTomorrow/calva/issues/1625)
+
 ## [2.0.307] - 2022-10-11
 
 - [Support user level `~/.config/calva/config.edn`](https://github.com/BetterThanTomorrow/calva/issues/1887)
@@ -30,7 +32,7 @@ Changes to Calva.
 ## [2.0.302] - 2022-09-18
 
 - [Show error message if loading file results in an error](https://github.com/BetterThanTomorrow/calva/issues/1767)
-- Document workaround for: [Allow sending a different project-root-uri during LSP initialize request](https://github.com/BetterThanTomorrow/calva/issues/1866) 
+- Document workaround for: [Allow sending a different project-root-uri during LSP initialize request](https://github.com/BetterThanTomorrow/calva/issues/1866)
 
 ## [2.0.301] - 2022-09-16
 
@@ -137,7 +139,7 @@ Changes to Calva.
 ## [2.0.280] - 2022-05-31
 
 - Fix: [Debugger decorations are not working properly](https://github.com/BetterThanTomorrow/calva/issues/1165)
-- Add some logging when Calva starts and finishes activating 
+- Add some logging when Calva starts and finishes activating
 
 ## [2.0.279] - 2022-05-30
 

--- a/src/config.ts
+++ b/src/config.ts
@@ -98,7 +98,7 @@ async function updateCalvaConfigFromEdn(uri?: vscode.Uri) {
     if (isDefined(uri)) {
       resolvedUri = uri;
     } else if (isDefined(configPath)) {
-      resolvedUri = vscode.Uri.file(configPath);
+      resolvedUri = configPath;
     } else {
       throw new Error('Expected a uri to be passed in or a config to exist at .calva/config.edn');
     }

--- a/src/extension-test/integration/suite/jack-in-test.ts
+++ b/src/extension-test/integration/suite/jack-in-test.ts
@@ -78,8 +78,7 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   await testUtil.openFile(testFilePath);
   testUtil.log(suite, 'test.clj opened');
 
-  const projectRootPath = await projectRoot.findClosestProjectRootPath();
-  const projetcRootUri = projectRootPath;
+  const projetcRootUri = await projectRoot.findClosestProjectRootPath();
   // Project type pre-select, qps = quickPickSingle
   const saveAs = `qps-${projetcRootUri.toString()}/jack-in-type`;
   await state.extensionContext.workspaceState.update(saveAs, projectType);

--- a/src/extension-test/integration/suite/jack-in-test.ts
+++ b/src/extension-test/integration/suite/jack-in-test.ts
@@ -78,9 +78,9 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   await testUtil.openFile(testFilePath);
   testUtil.log(suite, 'test.clj opened');
 
-  const projetcRootUri = await projectRoot.findClosestProjectRoot();
+  const projectRootUri = await projectRoot.findClosestProjectRoot();
   // Project type pre-select, qps = quickPickSingle
-  const saveAs = `qps-${projetcRootUri.toString()}/jack-in-type`;
+  const saveAs = `qps-${projectRootUri.toString()}/jack-in-type`;
   await state.extensionContext.workspaceState.update(saveAs, projectType);
 
   const res = commands.executeCommand(cmdId);

--- a/src/extension-test/integration/suite/jack-in-test.ts
+++ b/src/extension-test/integration/suite/jack-in-test.ts
@@ -79,7 +79,7 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   testUtil.log(suite, 'test.clj opened');
 
   const projectRootPath = await projectRoot.findClosestProjectRootPath();
-  const projetcRootUri = vscode.Uri.file(projectRootPath);
+  const projetcRootUri = projectRootPath;
   // Project type pre-select, qps = quickPickSingle
   const saveAs = `qps-${projetcRootUri.toString()}/jack-in-type`;
   await state.extensionContext.workspaceState.update(saveAs, projectType);

--- a/src/extension-test/integration/suite/jack-in-test.ts
+++ b/src/extension-test/integration/suite/jack-in-test.ts
@@ -78,7 +78,7 @@ async function startJackInProcedure(suite: string, cmdId: string, projectType: s
   await testUtil.openFile(testFilePath);
   testUtil.log(suite, 'test.clj opened');
 
-  const projetcRootUri = await projectRoot.findClosestProjectRootPath();
+  const projetcRootUri = await projectRoot.findClosestProjectRoot();
   // Project type pre-select, qps = quickPickSingle
   const saveAs = `qps-${projetcRootUri.toString()}/jack-in-type`;
   await state.extensionContext.workspaceState.update(saveAs, projectType);

--- a/src/file-switcher/file-switcher.ts
+++ b/src/file-switcher/file-switcher.ts
@@ -56,7 +56,7 @@ export async function toggleBetweenImplAndTest() {
   let fileToOpen = vscode.workspace.asRelativePath(filePath, false);
   if (fileToOpen[0] === '/') {
     // When connected over live share, the above will result in an absolute path,
-    // which causes that workspace.findFiles won't find the file. So we strip the
+    // which causes workspace.findFiles to not find the file. So we strip the
     // leading slash to prevent such scenario's.
     fileToOpen = fileToOpen.substring(1);
   }

--- a/src/file-switcher/file-switcher.ts
+++ b/src/file-switcher/file-switcher.ts
@@ -4,9 +4,10 @@ import * as util from './util';
 import * as projectRoot from '../project-root';
 import { getActiveTextEditor } from '../utilities';
 
-function openFile(file) {
+function openFile(file: string | vscode.Uri) {
+  const fileUri: vscode.Uri = file instanceof vscode.Uri ? file : vscode.Uri.file(file);
   return vscode.workspace
-    .openTextDocument(vscode.Uri.file(file))
+    .openTextDocument(fileUri)
     .then((doc) => vscode.window.showTextDocument(doc, { preserveFocus: false }));
 }
 
@@ -14,22 +15,19 @@ function showConfirmationDialog(text, button) {
   return vscode.window.showWarningMessage(text, { modal: true }, button);
 }
 
-async function createNewFile(dir, file) {
-  const uriDir = vscode.Uri.file(dir);
-  const uriFile = vscode.Uri.file(path.join(dir, file));
+async function createNewFile(dir: vscode.Uri, fileUri: vscode.Uri) {
+  await vscode.workspace.fs.createDirectory(dir);
   const ws = new vscode.WorkspaceEdit();
-
-  await vscode.workspace.fs.createDirectory(uriDir);
-  ws.createFile(uriFile);
+  ws.createFile(fileUri);
   await vscode.workspace.applyEdit(ws);
 }
 
-function askToCreateANewFile(dir, file) {
-  const filePath = path.join(dir, file);
-  return showConfirmationDialog(`Create ${filePath}?`, 'Create').then((answer) => {
+function askToCreateANewFile(dir: vscode.Uri, filename: string) {
+  const fileUri = vscode.Uri.joinPath(dir, filename);
+  return showConfirmationDialog(`Create ${fileUri.fsPath}?`, 'Create').then((answer) => {
     if (answer === 'Create') {
-      void createNewFile(dir, file).then(() => {
-        void openFile(filePath);
+      void createNewFile(dir, fileUri).then(() => {
+        void openFile(fileUri);
       });
     }
   });
@@ -38,7 +36,8 @@ function askToCreateANewFile(dir, file) {
 export async function toggleBetweenImplAndTest() {
   const activeFile = getActiveTextEditor();
   const openedFilename = activeFile.document.fileName;
-  const projectRootPath = await projectRoot.findClosestProjectRootPath();
+  const projectRootUri = await projectRoot.findClosestProjectRootPath();
+  const projectRootPath = projectRootUri.fsPath;
   const pathAfterRoot = openedFilename.replace(projectRootPath, '');
   const fullFileName = path.basename(openedFilename);
   const extension = path.extname(fullFileName);
@@ -54,13 +53,19 @@ export async function toggleBetweenImplAndTest() {
   const newFilename = util.getNewFilename(fileName, extension);
 
   const filePath = path.join(projectRootPath, sourcePath, newFilename);
-  const fileToOpen = vscode.workspace.asRelativePath(filePath, false);
+  let fileToOpen = vscode.workspace.asRelativePath(filePath, false);
+  if (fileToOpen[0] === '/') {
+    // When connected over live share, the above will result in an absolute path,
+    // which causes that workspace.findFiles won't find the file. So we strip the
+    // leading slash to prevent such scenario's.
+    fileToOpen = fileToOpen.substring(1);
+  }
 
   void vscode.workspace.findFiles(fileToOpen, projectRoot.excludePattern()).then((files) => {
     if (!files.length) {
-      void askToCreateANewFile(path.join(projectRootPath, sourcePath), newFilename);
+      void askToCreateANewFile(vscode.Uri.joinPath(projectRootUri, sourcePath), newFilename);
     } else {
-      const file = files[0].fsPath;
+      const file = files[0];
       void openFile(file);
     }
   });

--- a/src/file-switcher/file-switcher.ts
+++ b/src/file-switcher/file-switcher.ts
@@ -57,7 +57,7 @@ export async function toggleBetweenImplAndTest() {
   if (fileToOpen[0] === '/') {
     // When connected over live share, the above will result in an absolute path,
     // which causes workspace.findFiles to not find the file. So we strip the
-    // leading slash to prevent such scenario's.
+    // leading slash to prevent such scenarios.
     fileToOpen = fileToOpen.substring(1);
   }
 

--- a/src/file-switcher/file-switcher.ts
+++ b/src/file-switcher/file-switcher.ts
@@ -36,7 +36,7 @@ function askToCreateANewFile(dir: vscode.Uri, filename: string) {
 export async function toggleBetweenImplAndTest() {
   const activeFile = getActiveTextEditor();
   const openedFilename = activeFile.document.fileName;
-  const projectRootUri = await projectRoot.findClosestProjectRootPath();
+  const projectRootUri = await projectRoot.findClosestProjectRoot();
   const projectRootPath = projectRootUri.fsPath;
   const pathAfterRoot = openedFilename.replace(projectRootPath, '');
   const fullFileName = path.basename(openedFilename);

--- a/src/files-cache.ts
+++ b/src/files-cache.ts
@@ -21,13 +21,13 @@ function writeToCache(uri: vscode.Uri) {
 export const content = (path: string | undefined) => {
   const resolvedPath = state.resolvePath(path);
   if (resolvedPath) {
-    if (!filesCache.has(resolvedPath)) {
-      writeToCache(vscode.Uri.file(resolvedPath));
-      const filesWatcher = vscode.workspace.createFileSystemWatcher(resolvedPath);
+    if (!filesCache.has(resolvedPath.fsPath)) {
+      writeToCache(resolvedPath);
+      const filesWatcher = vscode.workspace.createFileSystemWatcher(resolvedPath.fsPath);
       filesWatcher.onDidChange(writeToCache);
       filesWatcher.onDidCreate(writeToCache);
       filesWatcher.onDidDelete((uri) => filesCache.delete(uri.fsPath));
     }
-    return filesCache.get(resolvedPath);
+    return filesCache.get(resolvedPath.fsPath);
   }
 };

--- a/src/project-root.ts
+++ b/src/project-root.ts
@@ -3,7 +3,7 @@ import * as util from './utilities';
 import * as config from './config';
 import * as path from 'path';
 
-export async function findProjectRootPaths() {
+export async function findProjectRoots(): Promise<vscode.Uri[]> {
   const projectFileNames: string[] = [
     'project.clj',
     'shadow-cljs.edn',
@@ -33,10 +33,10 @@ export function excludePattern(moreExcludes: string[] = []) {
   return `**/{${[...moreExcludes, ...config.getConfig().projectRootsSearchExclude].join(',')}}`;
 }
 
-export async function findClosestProjectRootPath(candidatePaths?: vscode.Uri[]) {
+export async function findClosestProjectRoot(candidatePaths?: vscode.Uri[]): Promise<vscode.Uri> {
   const doc = util.tryToGetDocument({});
   const docDir = doc && doc.uri ? doc.uri.with({ path: path.dirname(doc.uri.fsPath) }) : undefined;
-  candidatePaths = candidatePaths ?? (await findProjectRootPaths());
+  candidatePaths = candidatePaths ?? (await findProjectRoots());
   const closestRootPath = docDir
     ? candidatePaths
         .filter((u) => docDir.fsPath.startsWith(u.fsPath))
@@ -50,10 +50,10 @@ export async function findClosestProjectRootPath(candidatePaths?: vscode.Uri[]) 
   }
 }
 
-export async function pickProjectRootPath(
+export async function pickProjectRoot(
   candidatePaths: vscode.Uri[],
   closestRootPath: vscode.Uri
-) {
+): Promise<vscode.Uri> {
   if (closestRootPath !== undefined) {
     const pickedRootPath =
       candidatePaths.length < 2

--- a/src/project-root.ts
+++ b/src/project-root.ts
@@ -16,14 +16,14 @@ export async function findProjectRootPaths() {
   const projectFilesGlob = `**/{${projectFileNames.join(',')}}`;
   const excludeDirsGlob = excludePattern();
   const t0 = new Date().getTime();
-  const rootPaths: string[] = [];
+  const rootPaths: vscode.Uri[] = [];
   if (vscode.workspace.workspaceFolders?.length > 0) {
-    const wsRootPaths = vscode.workspace.workspaceFolders.map((f) => f.uri.fsPath);
+    const wsRootPaths = vscode.workspace.workspaceFolders.map((f) => f.uri);
     rootPaths.push(...wsRootPaths);
   }
   const candidateUris = await vscode.workspace.findFiles(projectFilesGlob, excludeDirsGlob, 10000);
   console.debug('glob took', new Date().getTime() - t0, 'ms');
-  const projectFilePaths = candidateUris.map((uri) => path.dirname(uri.fsPath));
+  const projectFilePaths = candidateUris.map((uri) => uri.with({ path: path.dirname(uri.fsPath) }));
   rootPaths.push(...projectFilePaths);
   const candidatePaths = [...new Set(rootPaths)].sort();
   return candidatePaths;
@@ -33,13 +33,13 @@ export function excludePattern(moreExcludes: string[] = []) {
   return `**/{${[...moreExcludes, ...config.getConfig().projectRootsSearchExclude].join(',')}}`;
 }
 
-export async function findClosestProjectRootPath(candidatePaths?: string[]) {
+export async function findClosestProjectRootPath(candidatePaths?: vscode.Uri[]) {
   const doc = util.tryToGetDocument({});
-  const docDir = doc && doc.uri ? path.dirname(doc.uri.fsPath) : undefined;
+  const docDir = doc && doc.uri ? doc.uri.with({ path: path.dirname(doc.uri.fsPath) }) : undefined;
   candidatePaths = candidatePaths ?? (await findProjectRootPaths());
   const closestRootPath = docDir
     ? candidatePaths
-        .filter((p) => docDir.startsWith(p))
+        .filter((u) => docDir.fsPath.startsWith(u.fsPath))
         .sort()
         .reverse()[0]
     : candidatePaths[0];
@@ -50,22 +50,24 @@ export async function findClosestProjectRootPath(candidatePaths?: string[]) {
   }
 }
 
-export async function pickProjectRootPath(candidatePaths: string[], closestRootPath: string) {
+export async function pickProjectRootPath(
+  candidatePaths: vscode.Uri[],
+  closestRootPath: vscode.Uri
+) {
   if (closestRootPath !== undefined) {
     const pickedRootPath =
       candidatePaths.length < 2
         ? undefined
         : await util.quickPickSingle({
             title: 'Project root',
-            values: candidatePaths,
-            default: closestRootPath,
+            values: [...new Set(candidatePaths.map((u) => u.fsPath))],
+            default: closestRootPath.fsPath,
             placeHolder: 'Multiple Clojure projects found. Please pick the one you want to use.',
             saveAs: `projectRoot`,
             autoSelect: true,
           });
-    const projectRootPath = candidatePaths.includes(pickedRootPath)
-      ? pickedRootPath
-      : closestRootPath;
-    return projectRootPath;
+    const filtered = candidatePaths.filter((u) => u.fsPath === pickedRootPath);
+    const pickedRootUri = filtered.length > 0 ? filtered[0] : undefined;
+    return pickedRootUri ?? closestRootPath;
   }
 }

--- a/src/state.ts
+++ b/src/state.ts
@@ -183,12 +183,17 @@ export async function initProjectDir(uri?: vscode.Uri): Promise<void> {
  * Tries to resolve absolute path in relation to project root
  * @param filePath - absolute or relative to the project
  */
-export function resolvePath(filePath?: string) {
+export function resolvePath(filePath?: string): vscode.Uri {
   const root = getProjectWsFolder();
-  if (filePath && path.isAbsolute(filePath)) {
-    return filePath;
+
+  if (root.uri.scheme !== 'file') {
+    return vscode.Uri.joinPath(root.uri, filePath);
   }
-  return filePath && root && path.resolve(root.uri.fsPath, filePath);
+
+  if (filePath && path.isAbsolute(filePath)) {
+    return vscode.Uri.file(filePath);
+  }
+  return filePath && root && vscode.Uri.file(path.resolve(root.uri.fsPath, filePath));
 }
 
 export { extensionContext, outputChannel, connectionLogChannel, analytics };

--- a/src/state.ts
+++ b/src/state.ts
@@ -168,7 +168,7 @@ export async function initProjectDir(uri?: vscode.Uri): Promise<void> {
       closestRootPath
     );
     if (projectRootPath !== undefined) {
-      setStateValue(PROJECT_DIR_KEY, projectRootPath);
+      setStateValue(PROJECT_DIR_KEY, projectRootPath.fsPath);
       setStateValue(PROJECT_DIR_URI_KEY, projectRootPath);
     } else {
       await setOrCreateNonProjectRoot(extensionContext, true);

--- a/src/state.ts
+++ b/src/state.ts
@@ -161,11 +161,11 @@ export async function initProjectDir(uri?: vscode.Uri): Promise<void> {
     setStateValue(PROJECT_DIR_KEY, path.resolve(uri.fsPath));
     setStateValue(PROJECT_DIR_URI_KEY, uri);
   } else {
-    const candidatePaths: vscode.Uri[] = await projectRoot.findProjectRootPaths();
-    const closestRootPath: vscode.Uri = await projectRoot.findClosestProjectRootPath(
+    const candidatePaths: vscode.Uri[] = await projectRoot.findProjectRoots();
+    const closestRootPath: vscode.Uri = await projectRoot.findClosestProjectRoot(
       candidatePaths
     );
-    const projectRootPath: vscode.Uri = await projectRoot.pickProjectRootPath(
+    const projectRootPath: vscode.Uri = await projectRoot.pickProjectRoot(
       candidatePaths,
       closestRootPath
     );

--- a/src/state.ts
+++ b/src/state.ts
@@ -161,12 +161,17 @@ export async function initProjectDir(uri?: vscode.Uri): Promise<void> {
     setStateValue(PROJECT_DIR_KEY, path.resolve(uri.fsPath));
     setStateValue(PROJECT_DIR_URI_KEY, uri);
   } else {
-    const candidatePaths = await projectRoot.findProjectRootPaths();
-    const closestRootPath = await projectRoot.findClosestProjectRootPath(candidatePaths);
-    const projectRootPath = await projectRoot.pickProjectRootPath(candidatePaths, closestRootPath);
+    const candidatePaths: vscode.Uri[] = await projectRoot.findProjectRootPaths();
+    const closestRootPath: vscode.Uri = await projectRoot.findClosestProjectRootPath(
+      candidatePaths
+    );
+    const projectRootPath: vscode.Uri = await projectRoot.pickProjectRootPath(
+      candidatePaths,
+      closestRootPath
+    );
     if (projectRootPath !== undefined) {
       setStateValue(PROJECT_DIR_KEY, projectRootPath);
-      setStateValue(PROJECT_DIR_URI_KEY, vscode.Uri.file(projectRootPath));
+      setStateValue(PROJECT_DIR_URI_KEY, projectRootPath);
     } else {
       await setOrCreateNonProjectRoot(extensionContext, true);
     }

--- a/src/state.ts
+++ b/src/state.ts
@@ -162,9 +162,7 @@ export async function initProjectDir(uri?: vscode.Uri): Promise<void> {
     setStateValue(PROJECT_DIR_URI_KEY, uri);
   } else {
     const candidatePaths: vscode.Uri[] = await projectRoot.findProjectRoots();
-    const closestRootPath: vscode.Uri = await projectRoot.findClosestProjectRoot(
-      candidatePaths
-    );
+    const closestRootPath: vscode.Uri = await projectRoot.findClosestProjectRoot(candidatePaths);
     const projectRootPath: vscode.Uri = await projectRoot.pickProjectRoot(
       candidatePaths,
       closestRootPath

